pkg/v1/tkg/test/tkgctl/shared/e2e_common.go

```diff
@@ -223,3 +269,69 @@ func E2ECommonSpec(ctx context.Context, inputGetter func() E2ECommonSpecInput) {
 		By("Test successful !")
 	})
 }
+
+func getNextAvailableTkrVersion(ctx context.Context, currentTkrVersion string, mngClient client.Client) (string, error) {
+	tkrList := runv1alpha3.TanzuKubernetesReleaseList{}
+	mngClient.List(ctx, &tkrList)
+	tkrs := tkrList.Items
+	availableTkrs := make(map[string]string)
+	for i := range tkrs {
+		compatible := ""
+		for _, condition := range tkrs[i].Status.Conditions {
+			if condition.Type == runv1alpha3.ConditionCompatible {
+				compatible = string(condition.Status)
+			}
+		}
+		labels := tkrs[i].Labels
+		activeStatus := "True"
+		if labels != nil {
+			if _, exists := labels[runv1alpha3.LabelDeactivated]; exists {
+				activeStatus = "False"
+			}
+		}
+		if !strings.EqualFold(compatible, "true") || !strings.EqualFold(activeStatus, "true") {
+			continue
+		}
+
+		compareResult, err := utils.CompareVMwareVersionStrings(currentTkrVersion, tkrs[i].Spec.Version)
+		if err != nil {
+			return "", err
+		}
+
+		if compareResult < 0 {
+			availableTkrs[tkrs[i].Spec.Version] = tkrs[i].Name
+		}
+
+	}
+	var tkrVersions []string
+	for version := range availableTkrs {
+		tkrVersions = append(tkrVersions, version)
+	}
+	if len(tkrVersions) < 1 {
+		return "", fmt.Errorf("no TKR version available for upgrade")
+	}
+	nextAvailableVersion := tkrVersions[0]
+	for _, newVersion := range tkrVersions {
+		compareResult, err := utils.CompareVMwareVersionStrings(nextAvailableVersion, newVersion)
```

Comment:
Ideally this comparison could have been done at the original loop to select the smallest available tkr in one go

Reply:
not possible, because we can't be sure that the original list is order. so we have to get a list of all compatibles first, then we look for the smallest one.

Reply:
changed to suggestiong thanks.